### DISCUSSION TITLE
Closing Momentum Budget using WENO advection scheme

### DISCUSSION THREAD

**Discussion by weigwu** [CliMA/Oceananigans.jl#3363] [Computational science!]
Hi all,

I am trying to close the momentum/divergence/vorticity budgets using nonhydrostatic model outputs. The model uses a WENO fifth-order scheme for advection. I have been using the following approach to calculate the advection terms for momentum. However, I suspect it may be an approximation of the "true" momentum flux. Is there a way to obtain the advection terms that align with WENO scheme?

Many thanks,
Roger

```Julia

u, v, w = simulation.model.velocities

uu_x=@at((Center, Center, Center), u*∂x(u)),
vu_y=@at((Center, Center, Center), v*∂y(u)),
wu_z=@at((Center, Center, Center), w*∂z(u)),

```

**Comment by glwagner** [comment]
Your suspicions are correct!

The tendency kernel functions are here:

https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl

for example advection is

https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl#L71

which you can compute via `KernelFunctionOperation` by writing

```julia
using Oceananigans.Advection: div_𝐯u

advection = model.advection
u, v, w = velocities = model.velocities
uu_x = KernelFunctionOperation{Face, Center, Center}(div_𝐯u, grid, advection, velocities, u)
```

the other components are

```julia
using Oceananigans.Advection: div_𝐯v, div_𝐯w
vu_y = KernelFunctionOperation{Center, Face, Center}(div_𝐯v, grid, advection, velocities, v)
wu_z = KernelFunctionOperation{Center, Center, Face}(div_𝐯w, grid, advection, velocities, w)
```

that will work provided your model doesn't have background velocity fields. If it does, you need to use `total_velocities` as in the kernel function.

Once you have the `KernelFunctionOperation`s you can pass them directly to output writers, or wrap them in `Field` to compute somewhere else.

Note that these advection flux divergences are located at the same place as the variables to which they refer (not at Center, Center, Center, unless it is the tracer advective flux divergence).

**Reply by zhihua-zheng** [reply]
Is it possible to calculate the vertical and horizontal advection terms separately in a similar way?

**Reply by glwagner** [reply]
Yes, for example the advection of `u` is implemented here:

https://github.com/CliMA/Oceananigans.jl/blob/615b2a854e86cc57f3e900eb463a7fe1fe890f1d/src/Advection/momentum_advection_operators.jl#L54-L58

So you can pick this apart to calculate either just the x-component of the _flux_ using `_advective_momentum_flux_Uu`. To calculate the contribution of `_advective_momentum_flux_Uu` to the flux divergence, you'll have to write a function like

```julia
@inline x_flux_divergence(i, j, k, grid, advection, U, u) = 1/Vᶠᶜᶜ(i, j, k, grid) * (δxᶠᵃᵃ(i, j, k, grid, _advective_momentum_flux_Uu, advection, U, u)
```

and then use `x_flux_divergence` with the `KernelFunctionOperation`.

**Reply by zhihua-zheng** [reply]
Thank you! That's helpful! But `∂x(U*u)` is not the same as `U*∂x(u)`, right? What if one wants to calculate the latter?

**Reply by glwagner** [reply]
You're free to compute `U*∂x(u)` of course! The question posed was how to calculate momentum advection using the "native" advection scheme, ie the same numerical scheme that's used to integrate the momentum equation. With `NonhydrostaticModel` we're always using the "flux form", ie something like `∂x(U*u)`, with appropriate reconstructions of `U` and `u` depending on the scheme. (Note that the `HydrostaticFreeSurfaceModel` also supports a "vector invariant" scheme, but this still does not use the advective form `U*∂x(u)`.)

**Reply by glwagner** [reply]
So just to complete the thought --- while one could try to come up with a scheme to compute `U*∂x(u)` based on WENO reconstruction, that scheme wouldn't be the one used natively by the model.

Are you interested in using (some kind of) WENO reconstruction to compute `U*∂x(u)` even if `NonhydrostaticModel` doesn't use such a scheme to integrate the momentum equation? (If so, why?)

**Reply by zhihua-zheng** [reply]
I see, the native flux form is what I should use to close the budget, but each component of the flux divergence is not exactly the advection in each direction, which makes it a bit unclear when I want to compare individual advection terms. 

Since diagnosing terms in advective form would involve a scheme that was not used by the model anyway, I guess that may be unnecessary.

**Reply by glwagner** [reply]
Yes, that's true. Though terms like `u*u` are indeed the "momentum flux", and there are three of them in the x-momentum equation: `u*u`, `v*u`, and `w*u`. So you can indeed decompose the flux divergence into three contributions if that is what you would like to do...
